Fonctions utilitaires en rapport avec les messages Discord

Steellgold opened the pull request:
- Close #211

Steellgold committed:
test(message.test.ts): add tests for isDiscordLink function
feat(message.util.ts): add isDiscordLink function to check if a link is a valid discord link

Steellgold committed:
refactor(resolve.cmd.ts): replace trustDiscordLinks with isDiscordLink function from message.util module

Bluzzi reviewed (changes requested):
Je me demande si le fichier utilitaire `message` n'aurait pas sa place au côté des fichiers utilitaires dans `/utils/function` ?

Unarray commented on src/utils/message/message.test.ts:
```diff
@@ -33,4 +33,30 @@ describe("msgParams function", () => {
   it("should ignore extra words when there are no placeholders", () => {
     expect(msgParams("Hello", ["Hello", "World"])).toBe("Hello");
   });
+});
+
+describe("isDiscordLink function", () => {
+  it("should return true when the link is a discord canary link", () => {
+    expect(isDiscordLink("https://canary.discord.com/channels/732251741999071303/1113731433123565589/1113731587570413588")).toBe(true);
+  });
+
+  it("should return true when the link is a discord ptb link", () => {
+    expect(isDiscordLink("https://ptb.discord.com/channels/732251741999071303/786216771723198514/1075749993631191110")).toBe(true);
+  });
+
+  it("should return true when the link is a discord link", () => {
+    expect(isDiscordLink("https://discord.com/channels/732251741999071303/786216771723198514/803532192793493544")).toBe(true);
+  });
+
+  it("should return false because is not valid", () => {
+    expect(isDiscordLink("https://discord.com/channels/732251741999071303")).toBe(false);
+  });
+
+  it("should return false because is litteraly not a discord link", () => {
+    expect(isDiscordLink("https://www.youtube.com/watch?v=dQw4w9WgXcQ")).toBe(false);
+  });
+
+  it("should return false when the link is not a discord link", () => {
+    expect(isDiscordLink("https://fakediscord.com/channels/732251741999071303/1113731433123565589/1113731587570413588")).toBe(false);
+  });
```
Ne serait-il pas plus judicieux de déplacer ces fonctions dans [`/utils/discord/`](https://github.com/Virtual-Royaume/Royaume-Discord-Bot/tree/main/src/utils/discord/) ?
Car si on déplace cette fonction dans [`/utils/function/`](https://github.com/Virtual-Royaume/Royaume-Discord-Bot/tree/main/src/utils/function/) il faudra aussi déplacer ce qu'il y a dans [`/utils/discord/`](https://github.com/Virtual-Royaume/Royaume-Discord-Bot/tree/main/src/utils/discord/) pour que ca soit un minimum cohérent.

Bluzzi replied:
Dans `/utils/discord` c'est destiné aux fonctions qui utilise directement l'API Discord, la ou `/utils/function` est fait pour des fonctions totalement native à JS et qui peuvent donc être testé de manière isolé.

Unarray replied:
Effectivement, je parlais surtout pour la future fonction de récupération d'un message

Steellgold committed:
feat(utils/message): add isDiscordLink and getMessageContentFromLink functions
test(utils/message): add tests for isDiscordLink, getMessageContentFromLink and msgParams functions

Steellgold commented:
![image](https://github.com/Virtual-Royaume/Royaume-Discord-Bot/assets/51505384/1b8628f5-cada-43c4-931d-70c6ca33e9e1)

Bluzzi committed:
Merge branch 'main' into fetch-message

Steellgold committed:
refactor(utils/message): rename isDiscordLink to containsDiscordLink and remove getMessageContentFromLink function
test(utils/message): add tests for containsDiscordLink function
feat(utils/message): export containsDiscordLink function instead of isDiscordLink function

Steellgold committed:
refactor(resolve.cmd.ts): rename isDiscordLink to containsDiscordLink for clarity and consistency with other functions

Steellgold committed:
feat(utils): add extractDiscordLink function to message.util.ts
test(utils): add tests for extractDiscordLink function in extract-discord-link.test.ts

Steellgold committed:
feat(message.util.ts): add discordLinkRegex to message.type.ts and import it to message.util.ts to avoid code duplication.

Steellgold committed:
refactor(message.util.ts): use match instead of test for containsDiscordLink function

Steellgold committed:
feat(utils): add getMessageFromLink function to get a message instance from a Discord message link
test(utils): add test for getMessageFromLink function to check if it returns an instance of Message

Steellgold committed:
test(get-message-instance.test.ts): add tests for PTB and Canary links and non-discord links

Bluzzi commented on src/utils/message/message.util.ts:
```diff
@@ -1,9 +1,44 @@
+};
+
+export const getMessageFromLink = async(link: string): Promise<Message<true> | null> => {
+  const ids = [...link.match(/(\d+)/g) ?? []];
+
+  if (ids.length !== 3) return null;
+
+  const guildId = ids[0];
+  const channelId = ids[1];
+  const messageId = ids[2];
```
La manière de parse est pourrait être amélioré :
```ts
const content = `
	Hello World!
 	https://canary.discord.com/channels/732251741999071303/786216771723198514/803532192793493544
  	are you ok ?
   	https://canary.discord.com/channels/732251741999071303/786216771723198514/803532192793493544
`;

const matchs = [...content.matchAll(/\/\d+\/\d+\/\d+/g)].map(element => {
  const [guildID, channelID, messageID] = [...element[0].matchAll(/\d+/g)].map(e => e[0]);
  
  return { guildID, channelID, messageID };
});

console.log(matchs);
```

Bluzzi commented on src/utils/message/_test/contains-discord-link.test.ts:
```diff
@@ -0,0 +1,32 @@
+describe("isDiscordLink function", () => {
+  it("should return true when the link is a discord canary link", () => {
+    expect(
+      containsDiscordLink("Salut check : https://canary.discord.com/channels/732251741999071303/1113731433123565589/1113731587570413588")
+    ).toBe(true);
```
En anglais les tests

Steellgold committed:
test(contains-discord-link.test.ts): update test descriptions and link examples

chore(contains-discord-link.test.ts): remove unnecessary test case

Steellgold committed:
chore(contains-discord-link.test.ts): remove contains-discord-link.test.ts file

chore(extract-discord-link.test.ts): update tests for extractDiscordLink function

chore(index.ts): remove containsDiscordLink from export

chore(message.type.ts): add DiscordMessage type

chore(message.util.ts): update extractDiscordLink function to return DiscordMessage or empty array

Steellgold committed:
fix(resolve.cmd.ts): change function name from containsDiscordLink to extractDiscordLink
fix(resolve.cmd.ts): change type of extracted message from array to DiscordMessage | null
fix(resolve.cmd.ts): change variable name from message to fetchedMessage

Arcoz0308 commented:
Que manque il a ce pr ? @Bluzzi

Bluzzi commented:
Certainement la résolution des reviews ouvertes, bien que des nouveaux commits ont vu le jour depuis celle-ci. On a pas eu de nouvelle de Gaëtan et il n'a pas actualisé le status des reviews.

Steellgold commented:
> Certainement la résolution des reviews ouvertes, bien que des nouveaux commits ont vu le jour depuis celle-ci. On a pas eu de nouvelle de Gaëtan et il n'a pas actualisé le status des reviews.

Je n'avais pas répondu / resolve mais j'avais commit les changements.
